File: test-errors/src/factories/internal/metadata/iterate_metadata_union.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.iterate_metadata_union = void 0;
const iterate_metadata_1 = require("./iterate_metadata");
const iterate_metadata_union = (checker) => (options) => (collection) => (errors) => (meta, type, explore) => {
    if (!type.isUnion())
        return false;
    type.types.forEach((t) => (0, iterate_metadata_1.iterate_metadata)(checker)(options)(collection)(errors)(meta, t, {
        ...explore,
        aliased: false,
    }));
    return true;
};
exports.iterate_metadata_union = iterate_metadata_union;

File: packages/typia/native/core/factories/internal/metadata/iterate_metadata_union.go
package metadata

func Iterate_metadata_union(props IMetadataIteratorProps) bool {
  if props.Type == nil || props.Type.IsUnion() == false {
    return false
  }
  members := props.Type.Types()
  before := props.Metadata.Size()
  for _, typ := range members {
    explore := props.Explore
    explore.Aliased = false
    Iterate_metadata(IMetadataIteratorProps{
      Options:     props.Options,
      Checker:     props.Checker,
      Components:  props.Components,
      Errors:      props.Errors,
      Metadata:    props.Metadata,
      Type:        typ,
      Explore:     explore,
      Intersected: props.Intersected,
      Unioned:     true,
      Prunable:    true,
    })
  }
  // Every member of a non-empty union pruned away to `never`, contributing no type
  // bucket — e.g. `Enum & { data: number }` distributes to
  // `("a" & { data }) | ("b" & { data })`, each member a non-object base meeting a
  // real-data object, which `is_never` drops. Left as an empty schema the union
  // would validate as accept-everything; mark it not-required so it renders as
  // `never` (rejecting every concrete value), matching a standalone `never`.
  //
  // `Nullable == false` is required: a `null` member sets `Nullable` without adding
  // a bucket, so `null | (string & { data })` (= `null`) reaches here with an empty
  // size. Forcing not-required there would also accept `undefined`, which `null`
  // must reject — so the guard fires only when the union carries no `null` either.
  if len(members) != 0 && props.Metadata.Any == false && props.Metadata.Nullable == false && props.Metadata.Size() == before {
    props.Metadata.Required = false
  }
  return true
}
